engine/src/ast/function_expr.rs

```diff
@@ -170,24 +179,49 @@ impl<'s> FunctionCallExpr<'s> {
     pub fn compile(self) -> CompiledValueExpr<'s> {
         let ty = self.get_type();
         let Self { function, args, .. } = self;
+        let map_each = args.get(0).and_then(|arg| arg.map_each_to());
         let args = args
             .into_iter()
             .map(FunctionCallArgExpr::compile)
             .collect::<Vec<_>>()
             .into_boxed_slice();
         let (mandatory_arg_count, optional_arg_count) = function.arg_count();
         let max_args = optional_arg_count.map_or_else(|| args.len(), |v| mandatory_arg_count + v);
-        CompiledValueExpr::new(move |ctx| {
-            match function.execute(&mut args.iter().map(|arg| arg.execute(ctx)).chain(
-                (args.len()..max_args).map(|index| Ok(function.default_value(index).unwrap())),
-            )) {
-                Some(value) => {
+        if let Some(map_each_type) = map_each {
```

Comment:
I think we need some comments here to explai what this code does, it's non-trivial :-)